### ISSUE TITLE
Resulting d.ts is invalid TS code: T_1 generic type argument

### ISSUE THREAD

**Issue by grissius** [microsoft/TypeScript#40362]
**TypeScript Version:**  4.0.2 (reproduced in 4.1.0-insiders.20200903 as well)

**Search Terms:**

template argument, overshadow argument, template argument aliasing

**Code**

```ts
// Any instance type
type Client = string

// Modified instance
type UpdatedClient<C> = C & {foo: number}

export const createClient = <
  D extends
    | (new (...args: any[]) => Client) // accept class
    | Record<string, new (...args: any[]) => Client> // or map of classes
>(
  clientDef: D
): D extends new (...args: any[]) => infer C
  ? UpdatedClient<C> // return instance
  : {
      [K in keyof D]: D[K] extends new (...args: any[]) => infer C // or map of instances respectively
        ? UpdatedClient<C>
        : never
    } => {
  return null as any
}
```

This is the simplest MWE I could find. Strangely enough, using a generic type instead of function does not cause the error.

**Expected behavior:**
I expect to find this type in the resulting `*.d.ts` to be a valid TS code and the type of the object type branch in return to be:
```ts
{ [K in keyof D]: D[K] extends new (...args: any[]) => infer C ? UpdatedClient<C> : never; }
```

**Actual behavior:**
`*.d.ts` is not a valid TS code according to the same version of TS. There is this new `C_1` which is a reference error.
```ts
{ [K in keyof D]: D[K] extends new (...args: any[]) => infer C ? UpdatedClient<C_1> : never; }
```

```
'C' is declared but its value is never read.
Cannot find name 'C_1'.
```

I think TS is probably trying to avoid C, since it assumes it's already reserved from `if` branch in the ternary type. Using different label does not produce the issue.

**Playground Link:** [link](https://www.typescriptlang.org/play?ts=4.0.2#code/PTAEEEDsE9QS0gZwC4ENIGMCmpnQA5YBQehoAwgDZxaTKgC8oKATggOZFEigCyA9gBM4AMxqD4SNJmKkcAVXyDUyLIKo06AHnIA+RhVAAyUAG8R-fgC5QkAK4BbAEZYWAXy5YAHvn4t6GPxSoBgsWCpYGrT0TFpEoKAAIqDeqpCCiPEJoAA+oAAUkFgA7gUAdBWoLOyINujQANoAugCUjPpRdG08qBjY+AGUqIiZ2bmgAEpYgSyCWqwcADS2JeWV1bWg9c1tDB3U0fo8fqAOqPig-CIhQyNYmbr5WRgHdIlYIjaJRC1fKV5pDIrUr5CplKo1OowHbtSQiVwULIAflAimUqnUr2QOiOYDCyDsLEgkhQ6GwWRspiyYwaAGlJKAANZYaBXJJNL50pr-QGIYFrcEbKGNVqwhDwliGY6Ss4XNkIUkyPlhRCEDDIOAANywlGg1LGoBRaIimM02L0+rGNiK2pY+rcsKpCXxhOJLzN72uwy2MCIbiAA)

**Related Issues:** Maybe #40302 and #36311
